store/pruning_block_store.go

```diff
@@ -0,0 +1,160 @@
+package store
+
+import (
+	"fmt"
+	"math"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+
+	"github.com/loomnetwork/loomchain/log"
+	"github.com/tendermint/tendermint/blockchain"
+	dbm "github.com/tendermint/tendermint/libs/db"
+)
+
+var (
+	calcBlockMetaPrefix = []byte("H:")
+	blockStoreKey       = []byte("blockStore")
+)
+
+func calcBlockMetaKey(height int64) []byte {
+	return []byte(fmt.Sprintf("H:%v", height))
+}
+
+func calcBlockPartKey(height int64, partIndex int) []byte {
+	return []byte(fmt.Sprintf("P:%v:%v", height, partIndex))
+}
+
+func calcBlockCommitKey(height int64) []byte {
+	return []byte(fmt.Sprintf("C:%v", height))
+}
+
+func calcSeenCommitKey(height int64) []byte {
+	return []byte(fmt.Sprintf("SC:%v", height))
+}
+
+type PruningBlockStore struct {
+	blockStoreDB       dbm.DB
+	prunedBlockStoreDB dbm.DB
+	*blockchain.BlockStore
+	cfg       *BlockStoreConfig
+	batchSize uint64
+}
+
+func NewPruningBlockStore(blockStoreConfig *BlockStoreConfig) *PruningBlockStore {
+	blockStoreDB := dbm.NewDB("blockstore", "leveldb", path.Join(blockStoreConfig.ChainDataPath, "data"))
+	prunedBlockStoreDB := dbm.NewDB("pruned_blockstore", "leveldb", path.Join(blockStoreConfig.ChainDataPath, "data"))
+	return &PruningBlockStore{
+		blockStoreDB:       blockStoreDB,
+		prunedBlockStoreDB: prunedBlockStoreDB,
+		BlockStore:         blockchain.NewBlockStore(blockStoreDB),
+		cfg:                blockStoreConfig,
+	}
+}
+
+func (bs *PruningBlockStore) Close() {
+	bs.blockStoreDB.Close()
+	bs.prunedBlockStoreDB.Close()
+}
+
+func (bs *PruningBlockStore) Height() int64 {
+	return blockchain.LoadBlockStoreStateJSON(bs.blockStoreDB).Height
+}
+
+func (bs *PruningBlockStore) Prune() error {
+	latestHeight := bs.Height()
```

Comment:
We also need to ensure that `latestHeight` is less than the current `app.db` height, otherwise the node may not be able to replay all the blocks from the current app height up to `latestHeight`.